### PR TITLE
KT-22971 Inspection to highlight and remove unnecessary explicit companion object references

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-22971

### CODE REVIEW

**Comment by udalov** [comment]
Please make sure that nothing is reported in
```
class C {
    companion object {
        fun foo() {}
    }
}

fun test() = C.Companion::foo  // should be green
```

File: idea/src/org/jetbrains/kotlin/idea/inspections/RedundantCompanionReferenceInspection.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            if (!DescriptorUtils.isCompanionObject(descriptor)) return
+
+            val parent = expression.getStrictParentOfType<KtDotQualifiedExpression>() ?: return
+            if (expression == parent.receiverExpression && expression.text == descriptor.containingDeclaration?.name?.asString()) return
+            if (parent.getStrictParentOfType<KtCallableReferenceExpression>() != null) return
+
+            holder.registerProblem(
+                expression,
+                "Redundant Companion reference",
+                ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
```
Comment (mglukhikh) [inline idea/src/org/jetbrains/kotlin/idea/inspections/RedundantCompanionReferenceInspection.kt]: LIKE_UNUSED_SYMBOL should be better here

File: idea/src/META-INF/plugin.xml [hunk 2]
```diff
@@ -2713,6 +2713,15 @@
                      language="kotlin"
     />
 
+    <localInspection implementationClass="org.jetbrains.kotlin.idea.inspections.RedundantCompanionReferenceInspection"
+                     displayName="Redundant Companion reference"
+                     groupPath="Kotlin"
+                     groupName="Style issues"
```
Comment (mglukhikh) [inline idea/src/META-INF/plugin.xml]: Redundant constructs should be better here

File: idea/src/META-INF/plugin.xml [hunk 3]
```diff
@@ -2713,6 +2713,15 @@
...
+                     enabledByDefault="true"
+                     level="WEAK WARNING"
```
Comment (mglukhikh) [inline idea/src/META-INF/plugin.xml]: I'd upgrade it to WARNING

File: idea/testData/inspectionsLocal/redundantCompanionReference/functionReference.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// PROBLEM: none
+class C {
+    companion object {
+        fun foo() {}
+    }
+}
+
+fun test() = C.<caret>Companion::foo
```
Comment (mglukhikh) [inline idea/testData/inspectionsLocal/redundantCompanionReference/functionReference.kt]: Please check also the following case: 
```
val obj = C.Companion // PROBLEM: none
```

File: idea/src/org/jetbrains/kotlin/idea/inspections/RedundantCompanionReferenceInspection.kt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+private class RemoveRedundantCompanionReferenceFix : LocalQuickFix {
+    override fun getName() = "Remove redundant Companion reference"
+
+    override fun getFamilyName() = name
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val expression = descriptor.psiElement as? KtReferenceExpression ?: return
+        val parent = expression.getStrictParentOfType<KtDotQualifiedExpression>() ?: return
```
Comment (asedunov) [inline idea/src/org/jetbrains/kotlin/idea/inspections/RedundantCompanionReferenceInspection.kt]: You can do it via rename: ```expression.mainReference.handleElementRename("")``` (look at ```KtSimpleNameReference.handleElementRename(String)```)

**Comment by mglukhikh** [comment]
Ok, I'll merge it, thank you

**Comment by mglukhikh** [comment]
Merged

**Pull request closed without merging**